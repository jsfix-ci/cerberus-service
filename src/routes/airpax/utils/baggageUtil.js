import * as pluralise from 'pluralise';
import { UNKNOWN_TEXT } from '../../../constants';

const getBaggageWeight = (baggage) => {
  if (!baggage?.weight) {
    return UNKNOWN_TEXT;
  }
  return baggage.weight;
};

const hasBaggage = (targetTask) => {
  return !!targetTask.movement.baggage;
};

const getBaggage = (targetTask) => {
  if (hasBaggage(targetTask)) {
    return targetTask.movement.baggage;
  }
  return null;
};

const getCheckedBags = (baggage, taskDetails = false) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% checked bag',
    '% checked bags',
    (taskDetails ? 'None' : 'No checked bags'),
  );
};

const getNumberOfCheckedBags = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return baggage.numberOfCheckedBags;
};

const toFormattedCheckedBagsCount = (baggage) => {
  if ((!baggage || !baggage?.numberOfCheckedBags) && baggage?.numberOfCheckedBags !== 0) {
    return UNKNOWN_TEXT;
  }
  return pluralise.withCount(
    baggage.numberOfCheckedBags,
    '% bag',
    '% bags',
    'None',
  );
};

const BaggageUtil = {
  get: getBaggage,
  weight: getBaggageWeight,
  checked: getCheckedBags,
  checkedCount: getNumberOfCheckedBags,
  formatCheckedCount: toFormattedCheckedBagsCount,
};

export default BaggageUtil;

export { getBaggageWeight, getBaggage, getCheckedBags, getNumberOfCheckedBags };
