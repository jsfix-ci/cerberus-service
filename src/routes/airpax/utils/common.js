import lookup from 'country-code-lookup';
import { UNKNOWN_TEXT } from '../../../constants';

const convertToIso3Code = (iso2Code) => {
  if (!iso2Code) {
    return UNKNOWN_TEXT;
  }
  if (!lookup.byIso(iso2Code)) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(iso2Code).iso3;
};

const Common = {
  iso3Code: convertToIso3Code,
};

export default Common;

export { convertToIso3Code };
