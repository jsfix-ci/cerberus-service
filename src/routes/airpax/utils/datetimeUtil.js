import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { UNKNOWN_TEXT } from '../../../constants';

dayjs.extend(utc);

const getDate = () => {
  return dayjs.utc().format();
};

const getFormattedDate = (date, dateFormat) => {
  if (!date) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc(date).format(dateFormat);
};

const toDateTimeList = (dateOne, dateTwo) => {
  return [dateOne, dateTwo];
};

const DateTimeUtil = {
  date: getDate,
  format: getFormattedDate,
  toList: toDateTimeList,
};

export default DateTimeUtil;

export { getDate, getFormattedDate, toDateTimeList };
