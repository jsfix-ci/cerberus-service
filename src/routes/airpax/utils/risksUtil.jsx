import React from 'react';
import * as pluralise from 'pluralise';
import { WARNING_CODES_MAPPING, NO_TEXT, YES_TEXT, CURRENTLY_UNAVAILABLE_TEXT } from '../../../constants';

const formatTargetIndicators = (targetingIndicators) => {
  if (targetingIndicators?.indicators?.length > 0) {
    const threatIndicatorList = targetingIndicators.indicators.map((threatIndicator) => {
      return threatIndicator.description;
    });
    return (
      <ul className="govuk-list item-list--bulleted">
        <li className="govuk-!-font-weight-bold govuk-!-font-size-16">{`${pluralise.withCount(threatIndicatorList.length, '% indicator', '% indicators')}`}</li>
        {threatIndicatorList.map((threat) => {
          return <li key={threat} className="threat-indicator-bullet govuk-!-font-size-16">{threat}</li>;
        })}
      </ul>
    );
  }
};

const hasTargetingIndicators = (risks) => {
  return !!risks?.targetingIndicators;
};

const getTargetingIndicators = (risks) => {
  if (hasTargetingIndicators(risks)) {
    return risks.targetingIndicators;
  }
  return null;
};

const hasRisk = (targetTask) => {
  return !!targetTask?.risks;
};

const getRisk = (targetTask) => {
  if (hasRisk(targetTask)) {
    return targetTask.risks;
  }
  return null;
};

const getSelectorWarning = (selector) => {
  let warning;
  let warningDetails;
  const warningStatus = selector.warning.status;
  if (warningStatus?.toLowerCase() === NO_TEXT.toLowerCase()) warning = 'No warnings';
  if (warningStatus?.toLowerCase() === CURRENTLY_UNAVAILABLE_TEXT.toLowerCase()) warning = 'Warnings currently unavailable';
  if (warningStatus?.toLowerCase() === YES_TEXT.toLowerCase()) {
    const warningTypes = selector.warning.types;
    const containsOther = warningTypes.indexOf('O') > -1;
    if (warningTypes.length > 0) {
      if (containsOther) {
        warningDetails = selector.warning.detail;
      }
      warning = warningTypes.map((w) => (w === 'O' ? warningDetails?.substring(0, 500) : WARNING_CODES_MAPPING[w])).join(', ');
    }
  }
  return warning;
};

const hasIndicatorMatches = (selector) => {
  return !!selector?.indicatorMatches;
};

const getIndicatorMatches = (selector) => {
  if (hasIndicatorMatches(selector)) {
    return selector.indicatorMatches;
  }
  return null;
};

const hasSelectorGroups = (version) => {
  return !!version?.risks?.matchedSelectorGroups;
};

const getSelectorGroups = (version) => {
  if (hasSelectorGroups(version)) {
    return version.risks.matchedSelectorGroups;
  }
  return null;
};

const hasMatchedRules = (version) => {
  return !!version?.risks?.matchedRules;
};

const getMatchedRules = (version) => {
  if (hasMatchedRules(version)) {
    return version.risks.matchedRules;
  }
  return null;
};

const hasHighestThreatLevel = (risks) => {
  return !!risks?.highestThreatLevel;
};

const getHighestThreatLevel = (risks) => {
  if (hasHighestThreatLevel(risks)) {
    return risks.highestThreatLevel;
  }
  return null;
};

const hasRiskMatchedSelectorGroups = (risks) => {
  return !!risks?.matchedSelectorGroups;
};

const getRiskMatchedSelectorGroups = (risks) => {
  if (hasRiskMatchedSelectorGroups(risks)) {
    return risks.matchedSelectorGroups;
  }
  return null;
};

const hasRiskMatchedRules = (risks) => {
  return !!risks?.matchedRules;
};

const getRiskMatchedRules = (risks) => {
  if (hasRiskMatchedRules(risks)) {
    return risks.matchedRules;
  }
  return null;
};

const extractRiskType = (risks, highestRisk) => {
  const riskType = [];
  if (highestRisk.type.toLowerCase() === 'selector') {
    const selectors = [getRiskMatchedSelectorGroups(risks).groups];
    selectors.forEach((selector) => {
      Object.values(selector).every((s) => {
        if (s.category.toLowerCase() === (highestRisk.value.toLowerCase())) {
          riskType.push(s.threatType);
        }
      });
    });
  } else {
    const rules = [getRiskMatchedRules(risks)];
    rules.forEach((rule) => {
      Object.values(rule).every((r) => {
        if (r.priority.toLowerCase() === (highestRisk.value.toLowerCase())) {
          riskType.push(r.abuseTypes[0]);
        }
      });
    });
  }
  return riskType[0];
};

const formatTargetRisk = (risks, highestThreatLevel) => {
  const risksRules = getRiskMatchedRules(risks).length + getRiskMatchedSelectorGroups(risks).totalNumberOfSelectors;
  if (highestThreatLevel) {
    const topRisk = extractRiskType(risks, highestThreatLevel);
    const count = risksRules > 0 && risksRules - 1;
    return `${topRisk} and ${pluralise.withCount(
      count,
      '% other rule',
      '% other rules',
    )}`;
  }
  return null;
};

const formatHighestThreatLevel = (targetTask) => {
  const risks = getRisk(targetTask);
  const highestThreatLevel = getHighestThreatLevel(risks);
  if (!highestThreatLevel) {
    return;
  }
  return (
    <h4 className="govuk-heading-s task-highest-risk">
      {highestThreatLevel.type.toLowerCase() === 'rule' ? '' : highestThreatLevel.type}
      {' '}
      <span className="govuk-tag govuk-tag--riskTier">
        {highestThreatLevel.value}
      </span>
      <span className="govuk-body task-risk-statement">
        {formatTargetRisk(risks, highestThreatLevel)}
      </span>
    </h4>
  );
};

const RisksUtil = {
  getRisks: getRisk,
  format: formatTargetIndicators,
  getIndicators: getTargetingIndicators,
  getWarning: getSelectorWarning,
  getMatches: getIndicatorMatches,
  getGroups: getSelectorGroups,
  getRules: getMatchedRules,
  getHighestThreat: getHighestThreatLevel,
  getMatchedSelectorGroups: getRiskMatchedSelectorGroups,
  getMatchedRules: getRiskMatchedRules,
  formatHighestThreat: formatHighestThreatLevel,
};

export default RisksUtil;

export {
  getRisk,
  formatTargetIndicators,
  getTargetingIndicators,
  getSelectorWarning,
  getIndicatorMatches,
  getSelectorGroups,
  getMatchedRules,
  getHighestThreatLevel,
  getRiskMatchedSelectorGroups,
  getRiskMatchedRules,
  formatHighestThreatLevel,
};
