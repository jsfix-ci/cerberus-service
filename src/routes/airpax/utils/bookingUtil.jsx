import React from 'react';
import lookup from 'country-code-lookup';
import { getFormattedDate, toDateTimeList } from './datetimeUtil';
import calculateTimeDifference from '../../../utils/calculateDatetimeDifference';

import {
  UNKNOWN_TEXT,
  SHORT_DATE_FORMAT_ALT,
  STANDARD_HOUR_MINUTE_FORMAT,
  STANDARD_CARD_EXPIRY_FORMAT,
} from '../../../constants';

const getAgentLocation = (agent) => {
  if (!agent?.location) {
    return UNKNOWN_TEXT;
  }
  return agent.location;
};

const getAgentIata = (agent) => {
  if (!agent?.iata) {
    return UNKNOWN_TEXT;
  }
  return agent.iata;
};

const getAgentName = (agent) => {
  if (!agent?.name) {
    return UNKNOWN_TEXT;
  }
  return agent.name;
};

const hasAgent = (booking) => {
  return !!booking?.agent;
};

const getAgent = (booking) => {
  if (hasAgent(booking)) {
    return booking.agent;
  }
  return null;
};

const hasPaymentCard = (payment) => {
  return !!payment?.card;
};

const getPaymentCard = (payment) => {
  if (hasPaymentCard(payment)) {
    return payment.card;
  }
  return null;
};

const getCardExpiry = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.expiry) {
    return getFormattedDate(paymentCard.expiry, STANDARD_CARD_EXPIRY_FORMAT);
  }
  return UNKNOWN_TEXT;
};

const getCardLastFourDigits = (payment) => {
  const paymentCard = getPaymentCard(payment);
  if (paymentCard?.number) {
    return paymentCard.number.substr(paymentCard.number.length - 4);
  }
  return UNKNOWN_TEXT;
};

const getPaymentAmount = (payment) => {
  if (!payment?.amount) {
    return UNKNOWN_TEXT;
  }
  return payment.amount;
};

const hasPayments = (booking) => {
  return !!booking?.payments?.length > 0;
};

const getPayments = (booking) => {
  if (hasPayments(booking)) {
    return booking.payments;
  }
  return null;
};

const getTicketType = (ticket) => {
  if (!ticket?.type) {
    return UNKNOWN_TEXT;
  }
  return ticket.type;
};

const getTicketNumber = (ticket) => {
  if (!ticket?.number) {
    return UNKNOWN_TEXT;
  }
  return ticket.number;
};

const hasTicket = (booking) => {
  return !!booking?.ticket;
};

const getTicket = (booking) => {
  if (hasTicket(booking)) {
    return booking.ticket;
  }
  return null;
};

const getBookingType = (booking) => {
  if (!booking?.type) {
    return UNKNOWN_TEXT;
  }
  return booking.type;
};

const getBookingCountryCode = (booking) => {
  if (!booking?.country) {
    return UNKNOWN_TEXT;
  }
  return booking.country;
};

const getBookingCountryName = (booking) => {
  const countryCode = getBookingCountryCode(booking);
  if (countryCode === UNKNOWN_TEXT) {
    return UNKNOWN_TEXT;
  }
  if (lookup.byIso(countryCode) === null) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(countryCode).country;
};

const toBookingDateText = (booking) => {
  if (!booking?.bookedAt) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(booking.bookedAt, SHORT_DATE_FORMAT_ALT);
};

const getBookingReference = (booking) => {
  if (!booking?.reference) {
    return UNKNOWN_TEXT;
  }
  return booking.reference;
};

const getCheckInAt = (booking) => {
  if (!booking?.checkInAt) {
    return UNKNOWN_TEXT;
  }
  return booking.checkInAt;
};

const toCheckInTimeText = (booking) => {
  const checkinPrefix = 'Check-in';
  if (!booking?.checkInAt) {
    return `${checkinPrefix} ${UNKNOWN_TEXT}`;
  }
  return `${checkinPrefix} ${getFormattedDate(booking.checkInAt, STANDARD_HOUR_MINUTE_FORMAT)}`;
};

const getBookedAt = (booking) => {
  return booking?.bookedAt;
};

const getBookedPriorToDeparture = (bookedAt, departureTime) => {
  const dateTimeList = toDateTimeList(bookedAt, departureTime);
  return calculateTimeDifference(dateTimeList);
};

const hasBooking = (targetTask) => {
  return !!targetTask?.movement?.booking;
};

const getBooking = (targetTask) => {
  if (hasBooking(targetTask)) {
    return targetTask.movement.booking;
  }
  return null;
};

const toPaymentsBlock = (booking) => {
  const payments = getPayments(booking);
  if (payments) {
    return payments.map((payment, index) => {
      return (
        <div key={index} className="bottom-border-thin govuk-!-margin-top-1 govuk-!-padding-bottom-1">
          <div className="font__bold">{getPaymentAmount(payment)}</div>
          <div className="font__light">
            Credit card ending {getCardLastFourDigits(payment)}, expiry {getCardExpiry(payment)}
          </div>
        </div>
      );
    });
  }
};

const BookingUtil = {
  get: getBooking,
  bookedAt: getBookedAt,
  checkInAt: getCheckInAt,
  bookedPrior: getBookedPriorToDeparture,
  toCheckInText: toCheckInTimeText,
  bookingRef: getBookingReference,
  toBookingText: toBookingDateText,
  countryCode: getBookingCountryCode,
  countryName: getBookingCountryName,
  bookingType: getBookingType,
  bookingTicket: getTicket,
  ticketNumber: getTicketNumber,
  ticketType: getTicketType,
  payments: getPayments,
  containsPayments: hasPayments,
  paymentAmount: getPaymentAmount,
  paymentCard: getPaymentCard,
  cardLastFourDigits: getCardLastFourDigits,
  cardExpiry: getCardExpiry,
  agent: getAgent,
  agentName: getAgentName,
  agentIata: getAgentIata,
  agentLocation: getAgentLocation,
  paymentsBlock: toPaymentsBlock,
};

export default BookingUtil;

export {
  toBookingDateText,
  getBookingReference,
  toCheckInTimeText,
  getBooking,
  getBookedAt,
  getCheckInAt,
  getBookedPriorToDeparture,
  getBookingCountryCode,
  getBookingCountryName,
  getBookingType,
  getTicket,
  getTicketNumber,
  getTicketType,
  getPayments,
  getPaymentAmount,
  getPaymentCard,
  getCardLastFourDigits,
  getCardExpiry,
  getAgent,
  getAgentName,
  getAgentIata,
  getAgentLocation,
  toPaymentsBlock,
};
