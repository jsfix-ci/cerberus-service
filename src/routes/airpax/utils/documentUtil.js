import moment from 'moment';
import lookup from 'country-code-lookup';
import {
  A_TITLE_CASE_TEXT,
  AN_TITLE_CASE_TEXT,
  AGO_TEXT,
  A_SMALL_TEXT,
  AN_SMALL_TEXT,
  UNKNOWN_TEXT,
  AFTER_TRAVEL_TEXT,
  BEFORE_TRAVEL_TEXT,
  SHORT_DATE_FORMAT_ALT,
} from '../../../constants';
import { formatField } from '../../../utils/formatField';
import { getFormattedDate } from './datetimeUtil';

const calculateExpiry = (passportExpiry, arrivalTime) => {
  const expiry = arrivalTime
    && passportExpiry !== 'Unknown'
    && `${arrivalTime},${moment(passportExpiry).format('YYYY-MM-DDTHH:mm:ss')}`;
  if (expiry) {
    return formatField('BOOKING_DATETIME', expiry)
      .split(', ')[1]
      .replace(BEFORE_TRAVEL_TEXT, AFTER_TRAVEL_TEXT)
      .replace(AGO_TEXT, BEFORE_TRAVEL_TEXT)
      .replace(A_SMALL_TEXT, A_TITLE_CASE_TEXT)
      .replace(AN_SMALL_TEXT, AN_TITLE_CASE_TEXT);
  }
  return UNKNOWN_TEXT;
};

const hasDocument = (person) => {
  return !!person?.document;
};

const getDocument = (person) => {
  if (hasDocument(person)) {
    return person.document;
  }
  return null;
};

const getDocumentExpiry = (document, taskDetails = false) => {
  const expiryPrefix = 'Expires';
  if (!document?.expiry) {
    return taskDetails ? UNKNOWN_TEXT : `${expiryPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}` : `${expiryPrefix} ${getFormattedDate(document?.expiry, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentValidity = (document, taskDetails = false) => {
  const validityPrefix = 'Valid from';
  if (!document?.validFrom) {
    return taskDetails ? UNKNOWN_TEXT : `${validityPrefix} ${UNKNOWN_TEXT}`;
  }
  return taskDetails ? `${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}` : `${validityPrefix} ${getFormattedDate(document?.validFrom, SHORT_DATE_FORMAT_ALT)}`;
};

const getDocumentType = (document) => {
  if (!document?.type) {
    return UNKNOWN_TEXT;
  }
  return document.type;
};

const getDocumentNumber = (document) => {
  if (!document?.number) {
    return UNKNOWN_TEXT;
  }
  return document.number;
};

const getDocumentName = (person) => {
  if (!person?.name) {
    return UNKNOWN_TEXT;
  }
  const firstNames = person?.name?.first || UNKNOWN_TEXT;
  const lastName = person?.name?.last || UNKNOWN_TEXT;
  return `${lastName.toUpperCase()}, ${firstNames}`;
};

// TODO finish implementation once data flows through
const getDocumentIdentification = (document) => {
  if (!document) {
    return UNKNOWN_TEXT;
  }
  return UNKNOWN_TEXT;
};

const getDocumentCountryOfIssue = (document, taskDetails = false) => {
  const countryOfIssuePrefix = 'Issued by';
  if (!document?.countryOfIssue) {
    return taskDetails ? UNKNOWN_TEXT : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }
  if (taskDetails) {
    return lookup.byIso(document.countryOfIssue) !== null
      ? `${lookup.byIso(document.countryOfIssue).country} (${document.countryOfIssue})`
      : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
  }

  return lookup.byIso(document.countryOfIssue) !== null
    ? `${countryOfIssuePrefix} ${document.countryOfIssue}`
    : `${countryOfIssuePrefix} ${UNKNOWN_TEXT}`;
};

const getDocumentCountryOfIssueCode = (document) => {
  if (!document?.countryOfIssue) {
    return UNKNOWN_TEXT;
  }
  return document.countryOfIssue;
};

const getDocumentNationality = (document, taskDetails = false) => {
  return getDocumentCountryOfIssue(document, taskDetails);
};

const getDocumentDOB = (document) => {
  if (!document?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(document?.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const DocumentUtil = {
  get: getDocument,
  docExpiry: getDocumentExpiry,
  docValidity: getDocumentValidity,
  docType: getDocumentType,
  docNumber: getDocumentNumber,
  docName: getDocumentName,
  docIdentification: getDocumentIdentification,
  docCountry: getDocumentCountryOfIssue,
  docCountryCode: getDocumentCountryOfIssueCode,
  docNationality: getDocumentNationality,
  docDOB: getDocumentDOB,
  calculateExpiry,
};

export default DocumentUtil;

export {
  getDocument,
  getDocumentExpiry,
  getDocumentValidity,
  getDocumentType,
  getDocumentNumber,
  getDocumentName,
  getDocumentIdentification,
  getDocumentCountryOfIssue,
  getDocumentCountryOfIssueCode,
  getDocumentNationality,
  getDocumentDOB,
  calculateExpiry,
};
