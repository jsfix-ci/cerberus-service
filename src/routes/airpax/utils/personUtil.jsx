import React from 'react';
import lookup from 'country-code-lookup';
import dayjs from 'dayjs';
import formatGender from '../../../utils/genderFormatter';
import { getFormattedDate } from './datetimeUtil';
import { SHORT_DATE_FORMAT_ALT, UNKNOWN_TEXT } from '../../../constants';

const getNationality = (person) => {
  if (!person?.nationality) {
    return UNKNOWN_TEXT;
  }
  return person.nationality;
};

const getCountryName = (person) => {
  if (!person?.nationality) {
    return UNKNOWN_TEXT;
  }
  return lookup.byIso(person.nationality).country;
};

const getDateOfBirth = (person) => {
  if (!person?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return getFormattedDate(person.dateOfBirth, SHORT_DATE_FORMAT_ALT);
};

const getAge = (person) => {
  if (!person?.dateOfBirth) {
    return UNKNOWN_TEXT;
  }
  return dayjs.utc().diff(dayjs(person.dateOfBirth), 'year');
};

const getGender = (person) => {
  if (!person) {
    return UNKNOWN_TEXT;
  }
  return formatGender(person.gender);
};

const getLastName = (person, capitalize = true) => {
  if (!person?.name?.last) {
    return capitalize ? UNKNOWN_TEXT.toUpperCase() : UNKNOWN_TEXT;
  }
  return capitalize ? person.name.last.toUpperCase() : person.name.last;
};

const getFirstName = (person) => {
  if (!person?.name?.first) {
    return UNKNOWN_TEXT;
  }
  return person.name.first;
};

const getFrequentFlyerNumber = (person) => {
  if (!person?.frequentFlyerNumber) {
    return UNKNOWN_TEXT;
  }
  return person.frequentFlyerNumber;
};

const hasSSRCodes = (person) => {
  return !!person?.ssrCodes?.length > 0;
};

const getSSRCodes = (person) => {
  if (!hasSSRCodes(person)) {
    return UNKNOWN_TEXT;
  }
  return person.ssrCodes.join(', ');
};

const toCoTravellers = (otherPersons) => {
  if (!otherPersons) {
    return <li className="govuk-!-font-weight-bold">None</li>;
  }
  const maxToDisplay = 4;
  const remaining = otherPersons.length > maxToDisplay ? otherPersons.length - maxToDisplay : 0;
  return otherPersons.map((person, index) => {
    if (index < maxToDisplay) {
      return (
        <li key={index} className="govuk-!-font-weight-bold">
          {getLastName(person)}, {getFirstName(person)}{(index !== maxToDisplay - 1)
                && (index !== otherPersons.length - 1) ? ',' : ''} {(remaining > 0 && index + 1 === maxToDisplay)
                  ? ` plus ${remaining} more` : ''}
        </li>
      );
    }
  });
};

const getAllPersons = (person, otherPersons) => {
  let allPersons = [];
  if (person) {
    allPersons.push(person);
  }
  if (otherPersons) {
    allPersons = allPersons.concat(otherPersons);
  }
  return allPersons;
};

const getTotalNumberOfPersons = (targetTask) => {
  if (!targetTask?.movement?.person) {
    return 0;
  }
  return targetTask.movement.otherPersons.length + 1;
};

const hasOtherPersons = (targetTask) => {
  return !!targetTask?.movement?.otherPersons?.length;
};

const getOtherPersons = (targetTask) => {
  if (hasOtherPersons(targetTask)) {
    return targetTask.movement.otherPersons;
  }
  return null;
};

const hasPerson = (targetTask) => {
  return !!targetTask?.movement?.person;
};

const getPerson = (targetTask) => {
  if (hasPerson(targetTask)) {
    return targetTask.movement.person;
  }
  return null;
};

const PersonUtil = {
  get: getPerson,
  getOthers: getOtherPersons,
  totalPersons: getTotalNumberOfPersons,
  toOthers: toCoTravellers,
  allPersons: getAllPersons,
  firstname: getFirstName,
  lastname: getLastName,
  gender: getGender,
  dob: getDateOfBirth,
  age: getAge,
  nationality: getNationality,
  countryName: getCountryName,
  frequentFlyerNumber: getFrequentFlyerNumber,
  ssrCodes: getSSRCodes,
};

export default PersonUtil;

export {
  getNationality,
  getCountryName,
  getDateOfBirth,
  getAge,
  getPerson,
  getGender,
  getFirstName,
  getLastName,
  toCoTravellers,
  getTotalNumberOfPersons,
  getOtherPersons,
  getFrequentFlyerNumber,
  getSSRCodes,
};
