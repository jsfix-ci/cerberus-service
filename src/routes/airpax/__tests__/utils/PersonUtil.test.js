import renderer from 'react-test-renderer';

import { UNKNOWN_TEXT } from '../../../../constants';
import { PersonUtil } from '../../utils';

describe('PersonUtil', () => {
  const person = {
    name: {
      first: 'Isaiah',
      last: 'Ford',
      full: 'Isaiah Ford',
    },
    role: 'PASSENGER',
    dateOfBirth: '1966-05-13T00:00:00Z',
    gender: 'M',
    nationality: 'GBR',
    document: null,
    ssrCodes: ['ABC'],
    frequentFlyerNumber: 123456,
  };

  const coTravellers = [
    {
      entitySearchUrl: null,
      name: {
        first: null,
        last: 'MAUSSER',
        full: 'MAUSSER',
      },
      role: 'CREW',
      dateOfBirth: '1970-04-14T00:00:00Z',
      gender: 'F',
      nationality: 'GB',
      document: {
        type: 'Passport',
        number: '1234567',
        countryOfIssue: 'FR',
        nationality: 'GB',
        validFrom: '1970-04-14T00:00:00Z',
        validTo: '1970-04-14T00:00:00Z',
        name: 'Miss MAUSSER MAUSSER',
        dateOfBirth: '1970-04-14T00:00:00Z',
      },
      movementStats: null,
      frequentFlyerNumber: null,
      ssrCodes: [
        'DOCS',
        'AUTH',
      ],
    },
    {
      entitySearchUrl: null,
      name: {
        first: 'SHARON',
        last: 'MAUSSER',
        full: 'SHARON MAUSSER',
      },
      role: 'CREW',
      dateOfBirth: null,
      gender: null,
      nationality: 'FR',
      document: null,
      movementStats: null,
      frequentFlyerNumber: '763381878A',
      ssrCodes: [
        'DOCS',
        'AUTH',
      ],
    },
    {
      entitySearchUrl: null,
      name: {
        first: null,
        last: null,
        full: null,
      },
      role: 'CREW',
      dateOfBirth: '1967-06-10T00:00:00Z',
      gender: 'M',
      nationality: 'GB',
      document: null,
      movementStats: null,
      frequentFlyerNumber: null,
      ssrCodes: [
        'DOCS',
        'AUTH',
      ],
    },
  ];

  it('should get a person object if present', () => {
    const targetTaskMin = {
      movement: {
        person: {
          name: {
            first: 'Isaiah',
            last: 'Ford',
            full: 'Isaiah Ford',
          },
          role: 'PASSENGER',
          dateOfBirth: '1966-05-13T00:00:00Z',
          gender: 'M',
          nationality: 'GBR',
          document: null,
          ssrCodes: ['ABC'],
          frequentFlyerNumber: 123456,
        },
      },
    };

    const output = PersonUtil.get(targetTaskMin);
    expect(output).toEqual(person);
  });

  it('should get other persons if present', () => {
    const targetTaskMin = {
      movement: {
        otherPersons: [
          {
            name: {
              first: 'Isaiah',
              last: 'Ford',
              full: 'Isaiah Ford',
            },
            role: 'PASSENGER',
            dateOfBirth: '1966-05-13T00:00:00Z',
            gender: 'M',
            nationality: 'GBR',
            document: null,
            ssrCodes: ['ABC'],
            frequentFlyerNumber: 123456,
          },
        ],
      },
    };

    const output = PersonUtil.getOthers(targetTaskMin);
    expect(output).toEqual([person]);
  });

  it('should return total number of people within a movement with 1 person', () => {
    const targetTaskMin = {
      movement: {
        person: {
          name: {
            first: 'Isaiah',
            last: 'Ford',
            full: 'Isaiah Ford',
          },
          role: 'PASSENGER',
          dateOfBirth: '1966-05-13T00:00:00Z',
          gender: 'M',
          nationality: 'GBR',
          document: null,
          ssrCodes: 'ABC',
          frequentFlyerNumber: 123456,
        },
        otherPersons: [],
      },
    };

    const output = PersonUtil.totalPersons(targetTaskMin);
    expect(output).toEqual(1);
  });

  it('should return total number of people within a movement with multiple persons', () => {
    const targetTaskMin = {
      movement: {
        person: {
          name: {
            first: 'Isaiah',
            last: 'Ford',
            full: 'Isaiah Ford',
          },
          role: 'PASSENGER',
          dateOfBirth: '1966-05-13T00:00:00Z',
          gender: 'M',
          nationality: 'GBR',
          document: null,
        },
        otherPersons: [
          {
            name: {
              first: 'John',
              last: 'Cheese',
              full: 'John Cheese',
            },
            role: 'PASSENGER',
            dateOfBirth: '1967-05-13T00:00:00Z',
            gender: 'M',
            nationality: 'GBR',
            document: null,
          },
        ],
      },
    };

    const output = PersonUtil.totalPersons(targetTaskMin);
    expect(output).toEqual(2);
  });

  it('should return nationality if present', () => {
    const output = PersonUtil.nationality(person);
    expect(output).toEqual(person.nationality);
  });

  it('should return dob if present', () => {
    const output = PersonUtil.dob(person);
    expect(output).toEqual('13 May 1966');
  });

  it('should show gender text when person is of gender male', () => {
    const output = PersonUtil.gender(person);
    expect(output).toEqual('Male');
  });

  it('should show gender text when person is of gender female', () => {
    person.gender = 'F';
    const output = PersonUtil.gender(person);
    expect(output).toEqual('Female');
  });

  it('should show unknown when person is of no specified gender', () => {
    person.gender = '';
    const output = PersonUtil.gender(person);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should show unknown when gender is null', () => {
    person.gender = null;
    const output = PersonUtil.gender(person);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should show unknown when gender is undefined', () => {
    person.gender = undefined;
    const output = PersonUtil.gender(person);
    expect(output).toEqual(UNKNOWN_TEXT);
  });

  it('should return first name if present', () => {
    const output = PersonUtil.firstname(person);
    expect(output).toEqual(person.name.first);
  });

  it('should return last name if present', () => {
    const output = PersonUtil.lastname(person);
    expect(output).toEqual(person.name.last.toUpperCase());
  });

  it('should return country name if nationality present', () => {
    const output = PersonUtil.countryName(person);
    expect(output).toEqual('United Kingdom');
  });

  it('should return frequent flyer number if present', () => {
    const output = PersonUtil.frequentFlyerNumber(person);
    expect(output).toEqual(person.frequentFlyerNumber);
  });

  it('should return SSR codes if present', () => {
    const output = PersonUtil.ssrCodes(person);
    expect(output).toEqual(person.ssrCodes.join(', '));
  });

  it('should calculate and return age if dob present', () => {
    const date = new Date();
    date.setFullYear(date.getFullYear() - 10);
    person.dateOfBirth = date.toISOString();

    const output = PersonUtil.age(person);
    expect(output).toEqual(10);
  });

  it('should return a formatted co-travellers block', () => {
    const otherPersons = [
      {
        name: {
          first: 'Isaiah',
          last: 'Ford',
          full: 'Isaiah Ford',
        },
        role: 'PASSENGER',
        dateOfBirth: '1966-05-13T00:00:00Z',
        gender: 'M',
        nationality: 'GBR',
        document: null,
      },
      {
        name: {
          first: 'Joe',
          last: 'Bloggs',
          full: 'Joe Bloggs',
        },
        role: 'PASSENGER',
        dateOfBirth: '1978-05-13T00:00:00Z',
        gender: 'M',
        nationality: 'FRA',
        document: null,
      },
    ];
    const tree = renderer.create(PersonUtil.toOthers(otherPersons)).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return a list of all persons within the movement', () => {
    const output = PersonUtil.allPersons(person, coTravellers);
    expect(output.length).toEqual(4);
  });
});
