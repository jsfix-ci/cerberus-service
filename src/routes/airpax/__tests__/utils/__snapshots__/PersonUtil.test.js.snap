// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonUtil should return a formatted co-travellers block 1`] = `
Array [
  <li
    className="govuk-!-font-weight-bold"
  >
    FORD
    , 
    Isaiah
    ,
     
    
  </li>,
  <li
    className="govuk-!-font-weight-bold"
  >
    BLOGGS
    , 
    Joe
    
     
    
  </li>,
]
`;
